src/github_runner_image_builder/builder.py

```diff
@@ -0,0 +1,635 @@
+        _resize_image(image_path=base_image_path)
+        logger.info("Mounting network block device.")
+        _mount_image_to_network_block_device(image_path=base_image_path)
+        logger.info("Replacing resolv.conf.")
+        _replace_mounted_resolv_conf()
+        logger.info("Resizing partitions.")
+        _resize_mount_partitions()
+        logger.info("Installing YQ from source.")
+        _install_yq()
+    except ImageBuilderBaseError as exc:
+        raise BuildImageError from exc
+
+    try:
+        logger.info("Setting up chroot environment.")
+        with ChrootContextManager(IMAGE_MOUNT_DIR):
+            # operator_libs_linux apt package uses dpkg -l and that does not work well with
+            # chroot env, hence use subprocess run.
+            output = subprocess.check_output(
+                ["/usr/bin/apt-get", "update", "-y"],
+                timeout=60 * 10,
+                env={"DEBIAN_FRONTEND": "noninteractive"},
+            )  # nosec: B603
+            logger.info("apt-get update out: %s", output)
+            output = subprocess.check_output(  # nosec: B603
+                [
+                    "/usr/bin/apt-get",
+                    "install",
+                    "-y",
+                    "--no-install-recommends",
+                    *IMAGE_DEFAULT_APT_PACKAGES,
+                ],
+                timeout=60 * 20,
+                env={"DEBIAN_FRONTEND": "noninteractive"},
+            )
+            logger.info("apt-get install out: %s", output)
+            _disable_unattended_upgrades()
+            _configure_system_users()
+            _install_yarn()
+    except ChrootBaseError as exc:
+        raise BuildImageError from exc
+    finally:
+        _clean_build_state()
+
+    try:
+        logger.info("Compressing image.")
+        _compress_image(base_image_path)
+    except ImageBuilderBaseError as exc:
+        raise BuildImageError from exc
+
+
+def _clean_build_state() -> None:
+    """Remove any artefacts left by previous build.
+
+    Raises:
+        CleanBuildStateError: if there was an error cleaning up the build state.
+    """
+    # The commands will fail if artefacts do not exist and hence there is no need to check the
+    # output of subprocess runs.
+    try:
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "dev")],
+            timeout=30,
+            check=False,
+        )  # nosec: B603
+        logger.info("umount dev out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "proc")],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )  # nosec: B603
+        logger.info("umount proc out: %s", output)
+        output = subprocess.run(
+            ["/usr/bin/umount", str(IMAGE_MOUNT_DIR / "sys")],
+            timeout=30,
+            check=False,
+            capture_output=True,
+        )  # nosec: B603
```

Comment:
Shouldn't this be managed by the chrootmanager ?

Reply:
If anything goes wrong while within the Chroot env, it will raise an exception and not go to `finally: _clean_build_state`. This means that any following image-builder calls will fail.
Doing a `clean_build_state` in every exception seems to be too verbose too.
What would you recommend in this situation?

Reply:
This does happen in the chroot env as well (`umount chroot_shared_dir` and `umount -l CHROOT_PATH / dev`) but this function exists to doubly make sure that the mounts are clean before and after the build.
What I could think of is to split `umount`s and `qemu-nbd disconnect`s and logically separate them.